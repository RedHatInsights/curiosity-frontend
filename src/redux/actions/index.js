import * as rhelActions from './rhelActions';
import * as userActions from './userActions';

const actions = {
  rhel: rhelActions,
  user: userActions
};

const reduxActions = { ...actions };

export { reduxActions as default, reduxActions, rhelActions, userActions };
