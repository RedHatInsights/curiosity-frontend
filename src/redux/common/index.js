import { reduxHelpers } from './reduxHelpers';

export { reduxHelpers as default, reduxHelpers };
