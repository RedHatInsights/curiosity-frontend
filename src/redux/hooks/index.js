import { reactReduxHooks } from './useReactRedux';

const storeHooks = {
  reactRedux: reactReduxHooks
};

export { storeHooks as default, storeHooks, reactReduxHooks };
