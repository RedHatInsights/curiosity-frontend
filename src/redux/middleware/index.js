import { createLogger } from 'redux-logger';
import promiseMiddleware from 'redux-promise-middleware';
import thunkMiddleware from 'redux-thunk';

const reduxMiddleware = [thunkMiddleware, promiseMiddleware];

if (process.env.NODE_ENV !== 'production' && process.env.REACT_APP_DEBUG_MIDDLEWARE === 'true') {
  reduxMiddleware.push(createLogger());
}

export { reduxMiddleware as default, reduxMiddleware, createLogger, promiseMiddleware, thunkMiddleware };
