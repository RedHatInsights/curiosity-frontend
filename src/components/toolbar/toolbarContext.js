import React from 'react';
import { ToolbarItem } from '@patternfly/react-core';
import { useProductQuery, useProductGraphTallyQuery, useProductToolbarConfig } from '../productView/productViewContext';
import { RHSM_API_QUERY_SET_TYPES } from '../../services/rhsm/rhsmConstants';
import { useOnSelect as useSelectCategoryOnSelect, toolbarFieldOptions } from './toolbarFieldSelectCategory';
import { useOnSelect as useBillingProviderOnSelect } from './toolbarFieldBillingProvider';
import { useOnSelect as useSlaOnSelect } from './toolbarFieldSla';
import { useOnSelect as useUsageOnSelect } from './toolbarFieldUsage';
import { SelectPosition } from '../form/select';
import { helpers } from '../../common/helpers';

/**
 * Return current values for categories/queries.
 *
 * @param {object} options
 * @param {Function} options.useProductQuery
 * @param {Function} options.useProductGraphTallyQuery
 * @returns {object}
 */
const useToolbarFieldQueries = ({
  useProductQuery: useAliasProductQuery = useProductQuery,
  useProductGraphTallyQuery: useAliasProductGraphTallyQuery = useProductGraphTallyQuery
} = {}) => {
  const {
    [RHSM_API_QUERY_SET_TYPES.BILLING_PROVIDER]: billingProvider,
    [RHSM_API_QUERY_SET_TYPES.SLA]: sla,
    [RHSM_API_QUERY_SET_TYPES.UOM]: uom,
    [RHSM_API_QUERY_SET_TYPES.USAGE]: usage
  } = useAliasProductQuery();
  const { [RHSM_API_QUERY_SET_TYPES.GRANULARITY]: granularity } = useAliasProductGraphTallyQuery();

  return {
    [RHSM_API_QUERY_SET_TYPES.BILLING_PROVIDER]: billingProvider,
    [RHSM_API_QUERY_SET_TYPES.GRANULARITY]: granularity,
    [RHSM_API_QUERY_SET_TYPES.SLA]: sla,
    [RHSM_API_QUERY_SET_TYPES.UOM]: uom,
    [RHSM_API_QUERY_SET_TYPES.USAGE]: usage
  };
};

/**
 * Clear a specific toolbar category.
 *
 * @param {object} options
 * @param {Function} options.useBillingProviderOnSelect
 * @param {Function} options.useSlaOnSelect
 * @param {Function} options.useUsageOnSelect
 * @returns {Function}
 */
const useToolbarFieldClear = ({
  useBillingProviderOnSelect: useAliasBillingProviderOnSelect = useBillingProviderOnSelect,
  useSlaOnSelect: useAliasSlaOnSelect = useSlaOnSelect,
  useUsageOnSelect: useAliasUsageOnSelect = useUsageOnSelect
} = {}) => {
  const billingOnSelect = useAliasBillingProviderOnSelect();
  const slaOnSelect = useAliasSlaOnSelect();
  const usageOnSelect = useAliasUsageOnSelect();

  return field => {
    switch (field) {
      case RHSM_API_QUERY_SET_TYPES.BILLING_PROVIDER:
        billingOnSelect();
        break;
      case RHSM_API_QUERY_SET_TYPES.SLA:
        slaOnSelect();
        break;
      case RHSM_API_QUERY_SET_TYPES.USAGE:
        usageOnSelect();
        break;
      default:
        break;
    }
  };
};

/**
 * Clear all available toolbar categories.
 *
 * @param {object} options
 * @param {Function} options.useProductQuery
 * @param {Function} options.useSelectCategoryOnSelect
 * @param {Function} options.useBillingProviderOnSelect
 * @param {Function} options.useSlaOnSelect
 * @param {Function} options.useUsageOnSelect
 * @returns {Function}
 */
const useToolbarFieldClearAll = ({
  useProductQuery: useAliasProductQuery = useProductQuery,
  useSelectCategoryOnSelect: useAliasSelectCategoryOnSelect = useSelectCategoryOnSelect,
  useBillingProviderOnSelect: useAliasBillingProviderOnSelect = useBillingProviderOnSelect,
  useSlaOnSelect: useAliasSlaOnSelect = useSlaOnSelect,
  useUsageOnSelect: useAliasUsageOnSelect = useUsageOnSelect
} = {}) => {
  const {
    [RHSM_API_QUERY_SET_TYPES.BILLING_PROVIDER]: billingProvider,
    [RHSM_API_QUERY_SET_TYPES.SLA]: sla,
    [RHSM_API_QUERY_SET_TYPES.USAGE]: usage
  } = useAliasProductQuery();
  const billingOnSelect = useAliasBillingProviderOnSelect();
  const slaOnSelect = useAliasSlaOnSelect();
  const usageOnSelect = useAliasUsageOnSelect();
  const selectCategoryOnSelect = useAliasSelectCategoryOnSelect();

  return hardFilterReset => {
    if (typeof billingProvider === 'string') {
      billingOnSelect();
    }

    if (typeof sla === 'string') {
      slaOnSelect();
    }

    if (typeof usage === 'string') {
      usageOnSelect();
    }

    if (hardFilterReset) {
      selectCategoryOnSelect();
    }
  };
};

/**
 * Return list of secondary toolbar fields for display.
 *
 * @param {object} options
 * @param {Array} options.categoryOptions
 * @param {Function} options.useProductToolbarConfig
 * @returns {Array}
 */
const useToolbarSecondaryFields = ({
  categoryOptions = toolbarFieldOptions,
  useProductToolbarConfig: useAliasProductToolbarConfig = useProductToolbarConfig
} = {}) => {
  const { secondaryFilters = [] } = useAliasProductToolbarConfig();

  return secondaryFilters.map(({ id, content }) => {
    const option = categoryOptions.find(({ value: categoryOptionValue }) => id === categoryOptionValue);
    const { component: OptionComponent } = option || {};

    return (
      (OptionComponent && (
        <ToolbarItem key={`option-${id}`}>
          <OptionComponent isFilter={false} position={SelectPosition.right} />
        </ToolbarItem>
      )) || <ToolbarItem key={helpers.generateId()}>{content}</ToolbarItem> ||
      null
    );
  });
};

const context = {
  useToolbarFieldClear,
  useToolbarFieldClearAll,
  useToolbarFieldQueries,
  useToolbarSecondaryFields
};

export {
  context as default,
  context,
  useToolbarFieldClear,
  useToolbarFieldClearAll,
  useToolbarFieldQueries,
  useToolbarSecondaryFields
};
