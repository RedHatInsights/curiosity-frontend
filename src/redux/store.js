import { createStore, applyMiddleware } from 'redux';
import { reduxMiddleware } from './middleware';
import { reduxReducers } from './reducers';

const store = createStore(reduxReducers, applyMiddleware(...reduxMiddleware));

export { store as default, store };
