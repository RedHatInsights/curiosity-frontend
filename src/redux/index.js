import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { withTranslation } from 'react-i18next';
import { store } from './store';
import { reduxActions } from './actions';
import { reduxReducers } from './reducers';
import { reduxSelectors } from './selectors';
import { reduxTypes } from './types';
import { helpers } from '../common/helpers';

const connectTranslate = (mapStateToProps, mapDispatchToProps) => component =>
  connect(
    mapStateToProps,
    mapDispatchToProps
  )((!helpers.TEST_MODE && withTranslation()(component)) || component);

const connectRouterTranslate = (mapStateToProps, mapDispatchToProps) => component =>
  withRouter(connectTranslate(mapStateToProps, mapDispatchToProps)(component));

const connectRouter = (mapStateToProps, mapDispatchToProps) => component =>
  withRouter(
    connect(
      mapStateToProps,
      mapDispatchToProps
    )(component)
  );

export {
  connect,
  connectRouter,
  connectRouterTranslate,
  connectTranslate,
  reduxActions,
  reduxReducers,
  reduxSelectors,
  reduxTypes,
  store
};
