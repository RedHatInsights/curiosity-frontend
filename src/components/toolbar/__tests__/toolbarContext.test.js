import { context, useToolbarFieldClear, useToolbarFieldClearAll, useToolbarFields } from '../toolbarContext';
import { store } from '../../../redux/store';
import { RHSM_API_QUERY_SET_TYPES as RHSM_API_QUERY_TYPES } from '../../../services/rhsm/rhsmConstants';

describe('ToolbarContext', () => {
  let mockDispatch;

  beforeEach(() => {
    mockDispatch = jest.spyOn(store, 'dispatch').mockImplementation((type, data) => ({ type, data }));
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should return specific properties', () => {
    expect(context).toMatchSnapshot('specific properties');
  });

  it('should apply a hook for clearing a toolbar field through redux', async () => {
    const { result: onClearField } = await renderHook(() => useToolbarFieldClear());
    onClearField('lorem');

    onClearField(RHSM_API_QUERY_TYPES.BILLING_PROVIDER);

    onClearField(RHSM_API_QUERY_TYPES.CATEGORY);

    onClearField(RHSM_API_QUERY_TYPES.SLA);

    onClearField(RHSM_API_QUERY_TYPES.USAGE);

    onClearField(RHSM_API_QUERY_TYPES.VARIANT);

    expect(mockDispatch.mock.calls).toMatchSnapshot('clear single field');
  });

  it('should apply a hook for clearing specific active toolbar fields through redux', async () => {
    const { result: onClearAllFields } = await renderHook(() =>
      useToolbarFieldClearAll({
        useProductQuery: () => ({
          lorem: 'ipsum',
          [RHSM_API_QUERY_TYPES.BILLING_PROVIDER]: 'testBillingProvider',
          [RHSM_API_QUERY_TYPES.CATEGORY]: 'testCategory',
          [RHSM_API_QUERY_TYPES.SLA]: 'testSla',
          [RHSM_API_QUERY_TYPES.USAGE]: 'testUsage',
          [RHSM_API_QUERY_TYPES.METRIC_ID]: 'testUomNowMetricId'
        })
      })
    );

    onClearAllFields();

    onClearAllFields(true);

    expect(mockDispatch.mock.calls).toMatchSnapshot('clear all fields');
  });

  it('should apply a hook for retrieving secondary toolbar field components', async () => {
    const { result: secondaryFields } = await renderHook(() =>
      useToolbarFields({
        useProductToolbarConfig: () => ({
          filters: [
            {
              id: 'rangedMonthly',
              isSecondary: true
            }
          ]
        })
      })
    );
    expect(secondaryFields).toMatchSnapshot('secondary field, rangedMonthly');
  });
});
