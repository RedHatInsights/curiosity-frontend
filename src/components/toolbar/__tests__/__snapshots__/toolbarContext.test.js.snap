// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToolbarContext should apply a hook for clearing a toolbar field through redux: clear single field 1`] = `
[
  [
    [
      {
        "type": "SET_QUERY_RESET_INVENTORY_LIST",
        "viewId": undefined,
      },
      {
        "sla": null,
        "type": "SET_QUERY_RHSM_sla",
        "viewId": undefined,
      },
    ],
  ],
  [
    [
      {
        "type": "SET_QUERY_RESET_INVENTORY_LIST",
        "viewId": undefined,
      },
      {
        "type": "SET_QUERY_RHSM_usage",
        "usage": null,
        "viewId": undefined,
      },
    ],
  ],
  [
    [
      {
        "type": "SET_QUERY_RESET_INVENTORY_LIST",
        "viewId": undefined,
      },
      {
        "billing_provider": null,
        "type": "SET_QUERY_RHSM_billing_provider",
        "viewId": undefined,
      },
    ],
  ],
]
`;

exports[`ToolbarContext should apply a hook for clearing specific active toolbar fields through redux: clear all fields 1`] = `
[
  [
    [
      {
        "type": "SET_QUERY_RESET_INVENTORY_LIST",
        "viewId": undefined,
      },
      {
        "billing_provider": null,
        "type": "SET_QUERY_RHSM_billing_provider",
        "viewId": undefined,
      },
    ],
  ],
  [
    [
      {
        "type": "SET_QUERY_RESET_INVENTORY_LIST",
        "viewId": undefined,
      },
      {
        "sla": null,
        "type": "SET_QUERY_RHSM_sla",
        "viewId": undefined,
      },
    ],
  ],
  [
    [
      {
        "type": "SET_QUERY_RESET_INVENTORY_LIST",
        "viewId": undefined,
      },
      {
        "type": "SET_QUERY_RHSM_usage",
        "usage": null,
        "viewId": undefined,
      },
    ],
  ],
  [
    [
      {
        "type": "SET_QUERY_RESET_INVENTORY_LIST",
        "viewId": undefined,
      },
      {
        "billing_provider": null,
        "type": "SET_QUERY_RHSM_billing_provider",
        "viewId": undefined,
      },
    ],
  ],
  [
    [
      {
        "type": "SET_QUERY_RESET_INVENTORY_LIST",
        "viewId": undefined,
      },
      {
        "sla": null,
        "type": "SET_QUERY_RHSM_sla",
        "viewId": undefined,
      },
    ],
  ],
  [
    [
      {
        "type": "SET_QUERY_RESET_INVENTORY_LIST",
        "viewId": undefined,
      },
      {
        "type": "SET_QUERY_RHSM_usage",
        "usage": null,
        "viewId": undefined,
      },
    ],
  ],
  [
    [
      {
        "currentFilter": null,
        "type": "SET_FILTER_TYPE",
        "viewId": undefined,
      },
    ],
  ],
]
`;

exports[`ToolbarContext should apply a hook for retrieving api queries specific to toolbar: query 1`] = `
{
  "billing_provider": "testProvider",
  "granularity": "testGranularity",
  "sla": "testSla",
  "uom": undefined,
  "usage": "testUsage",
}
`;

exports[`ToolbarContext should apply a hook for retrieving secondary toolbar field components: secondary field, rangedMonthly 1`] = `
[
  <ToolbarItem>
    <RangedMonthly
      isFilter={false}
      position="right"
    />
  </ToolbarItem>,
]
`;

exports[`ToolbarContext should return specific properties: specific properties 1`] = `
{
  "useToolbarFieldClear": [Function],
  "useToolbarFieldClearAll": [Function],
  "useToolbarFieldQueries": [Function],
  "useToolbarSecondaryFields": [Function],
}
`;
